// lds-rc/base.js

let BarTitle = require("./src/bar/title").default;
let BarNormal = require("./src/bar/normal").default;
let DialogNormal = require("./src/dialog/normal").default;
let SwitchNormal = require('./src/switch/normal').default;
let PickerColorBase = require('./src/picker/color/base').default;
let PickerColorColour = require('./src/picker/color/colour').default;
let PickerColorTemperature = require('./src/picker/color/temperature').default;
let SliderBrightness = require('./src/slider/brightness').default;
let LayoutSideTop = require('./src/layout/side/top').default;
let LayoutSideRight = require('./src/layout/side/right').default;
let LayoutSideBottom = require('./src/layout/side/bottom').default;
let LayoutSideLeft = require('./src/layout/side/left').default;
let LayoutNavTop = require('./src/layout/nav/top').default;
let LayoutNavRight = require('./src/layout/nav/right').default;
let LayoutNavBottom = require('./src/layout/nav/bottom').default;
let LayoutNavLeft = require('./src/layout/nav/left').default;
let LayoutMultilayer = require('./src/layout/multilayer').default;
let LayoutNavigation = require('./src/layout/navigation').default;
let DeviceCard = require('./src/card').default;

//客户定制类
let _ua = navigator.userAgent;
if(_ua.indexOf('com.a023.smarthome') !== -1){
	//todo
}else{
	//todo
}

export {
	BarTitle,
	BarNormal,
	DialogNormal,
	SwitchNormal,
	PickerColorBase,
	PickerColorColour,
	PickerColorTemperature,
	SliderBrightness,
	LayoutSideTop,
	LayoutSideRight,
	LayoutSideBottom,
	LayoutSideLeft,
	LayoutNavTop,
	LayoutNavRight,
	LayoutNavBottom,
	LayoutNavLeft,
	LayoutMultilayer,
	LayoutNavigation,
	DeviceCard
};
