// lds-rc/src/layout/nav/base.js
import React, { Component } from 'react';
import PropTypes from 'prop-types'; 
import ReactCSSTransitionGroup from "react-addons-css-transition-group";
import { Route, Link, NavLink, Switch } from 'react-router-dom';
import { matchPath } from 'react-router';

export default class LayoutNavigation extends Component {
  constructor(props, options) {
    super(props);
    this.state = Object.assign({
    	transitionName: "nav",
    	navTpl: '(item) => `<span class="icon ${item.iconClass}"></span><span class="title">${item.title}</span>`'
    }, props, options);
  }
  
  componentWillReceiveProps(nextProps){
  	//this.setState(nextProps);
  }
  
  handleNavClick(route, event){
  	let { location, history, children } = this.props;
  	if(!route.component){
  		location.state = {
  			view: "keep"
  		};
  	}
  }
  
  render() {
  	let { location, history, children } = this.props;
  	let { contentClass, contentStyle, navClass, navStyle, navLinkStyle, navTpl, contentKeep, navKeep } = this.state;
  	let pathname = location.pathname;
		let fnNavTpl = eval.call(null, navTpl);
    let isMatch = false;
    if(!(children instanceof Array)){
      children = [children];
    }
    for(let i=0; i<children.length; i++){
    	let item = children[i].props;
    	if(matchPath(pathname, item)){
    		isMatch = true;
    	}
    }
    let myLocation = Object.assign({}, location);
    if(isMatch){
    	this.myPathName = pathname;
    }else{
    	myLocation.pathname = this.myPathName;
    }
    let styles = {
    	content: Object.assign({
    		width: "100%",
    		display: ((contentKeep || isMatch) ? 'block' : 'none')
    	}, contentStyle),
    	nav: Object.assign({
    		link: {
    			display: "inline-block",
    			width: (100 / children.length) + "%"
    		},
	    	position: "fixed",
	    	zIndex: 101,
	    	width: "100%",
				textAlign: "center"
    	}, navStyle),
    	navLink: Object.assign({
  			display: "inline-block",
  			width: (100 / children.length) + "%"
    	}, navLinkStyle)
    }
		
    return (
			<React.Fragment>
				<div className={contentClass} style={styles.content}>
					<Switch location={contentKeep ? myLocation : location}>
						{children}
					</Switch>
				</div>
				<div className={navClass} style={styles.nav}>
				  {children.map((item, index) => 
        		<NavLink to={item.props.path} key={index} 
        			{...{replace:item.props.replace, exact:item.props.exact, strict:item.props.strict}}
        			onClick={this.handleNavClick.bind(this, item.props)}
        			location={navKeep ? myLocation : location}
        			activeClassName="active" style={styles.navLink}
        			dangerouslySetInnerHTML={{__html: fnNavTpl(item.props)}}>
    				</NavLink>
          )}
			  </div>
			</React.Fragment>
    );
  }
}

LayoutNavigation.propTypes = {
	location: PropTypes.object.isRequired,
	history: PropTypes.object.isRequired,
	children: PropTypes.array.isRequired,
	transitionName: PropTypes.string,
	navTpl: PropTypes.string,
  contentClass: PropTypes.string,
  contentStyle: PropTypes.object,
  navClass: PropTypes.string,
  navStyle: PropTypes.object,
  navLinkStyle: PropTypes.object
}
