// lds-rc/src/other/index.js
import React, { Component } from 'react';
import { BrowserRouter as Router, Route, NavLink, Link, Switch } from 'react-router-dom';
import helper from '../public/helper';
import DeviceCard from '../card';
import { Style, Lang } from './resource';

class Other extends Component {
  constructor(props) {
    super(props);
    this.state = {
			attr: {
				OnOff: 1,
				PowerLow: 1,
			},
			isLongPress: false,
    };
  }
	
  render() {
  	let {location, history} = this.props;
  	let props = {location, history};
		let myLocation = Object.assign({}, location);

		const clsName = {
      cardIcon: 'devices-icon',
      battery: 'battery',
      switch: 'switch-onoff',
      close: 'close',
    };
    return (
      <div className="other default">
	      <h2 style={{margin: "20px auto 10px"}}>设备列表卡片</h2>
				<div style={{textAlign: 'center'}}>
					<DeviceCard
						devId={'1'}
						devType={'Light_ColorTemperature'}
						name={'CCT灯'}
						online={true}
						icon={'lighting'}
						statusDesc={'Online'}
						roomId={'1'}
						roomName={'Bedroom'}
						attr={this.state.attr}
						cls={clsName}
						hasOnOffSwitch={true}
						hasLowPower={true}
						isLongPress={this.state.isLongPress}
						onLongPress={() => {
							this.setState({
								isLongPress: !this.state.isLongPress
							})
						}}
						onOnOff={() => {
							this.setState({
								attr: {
									OnOff: (!this.state.attr.OnOff) - 0,
									PowerLow: 1,
								}
							})
						}}
					/>
				</div>
      </div>
    );
  }
}

export default Other;
