// lds-rc/src/layout/side/base.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Route, Link, Switch } from 'react-router-dom';
import ReactCSSTransitionGroup from "react-addons-css-transition-group";
import { matchPath } from 'react-router';

export default class LayoutSideBase extends Component {
  constructor(props, options) {
  	super(props);
    this.state = Object.assign({
    	transitionName: "side"
    }, props, options);
    this.handleClickBackdrop = this.handleClickBackdrop.bind(this);
	}
  
  componentWillReceiveProps(nextProps){
  	//this.setState(nextProps);
  }
    
  handleClickBackdrop(event){
  	let { onClickBackdrop } = this.state;
  	onClickBackdrop && onClickBackdrop(event);
  }
  
  render() {
  	let { location, history, children } = this.props;
  	let { panelClass, panelStyle, backdropStyle, transitionName } = this.state;
  	let pathname = location.pathname;
    let isMatch = false;
    if(!(children instanceof Array)){
      children = [children];
    }
    for(let i=0; i<children.length; i++){
    	let item = children[i].props;
    	if(matchPath(pathname, item)){
    		isMatch = true;
    	}
    }
    let styles = {
    	backdrop: Object.assign({
    		position: "absolute",
    		display: (isMatch ? 'block' : 'none'),
			  width: "100vw",
			  height: "100vh",
			  top: 0,
			  left: 0,
				zIndex: 200
    	}, backdropStyle),
    	panel: Object.assign({
    		position: "absolute",
    		display: (isMatch ? 'block' : 'none'),
			  width: "100vw",
				zIndex: 201,
				overflow: "hidden"
    	}, panelStyle)
    }
  	let direction = this.direction[history.action];
    
		return (
			<React.Fragment>
      	{(styles.backdrop.background) && <div style={styles.backdrop} onClick={this.handleClickBackdrop}></div>}
				<ReactCSSTransitionGroup
	        transitionName={transitionName}
	        component="div"
			    className={direction}
	        transitionEnterTimeout={300}
	        transitionLeaveTimeout={300}>
					<div key={pathname} style={styles.panel}>
						<Switch location={location}>
							{children}
						</Switch>
			  	</div>
	      </ReactCSSTransitionGroup>
      </React.Fragment>
   	)
  }
}

LayoutSideBase.propTypes = {
	location: PropTypes.object.isRequired,
	history: PropTypes.object.isRequired,
	children: PropTypes.array.isRequired,
	transitionName: PropTypes.string,
	backdrop: PropTypes.string,
  panelClass: PropTypes.string
}
