// lds-rc/src/card/deviceCard.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './default/style.css';

const propTypes = {
  devId: PropTypes.string.isRequired,
  devType: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  online: PropTypes.bool.isRequired,
  icon: PropTypes.string.isRequired,
  hasOnOffSwitch: PropTypes.bool.isRequired, // 是否有可控开关
  hasLowPower: PropTypes.bool.isRequired, // 是否有低电量警报
  statusDesc: PropTypes.string.isRequired,
  roomId: PropTypes.string,
  roomName: PropTypes.string.isRequired,
  attr: PropTypes.object.isRequired,
  isLongPress: PropTypes.bool.isRequired,
  cls: PropTypes.object,
  onOnOff: PropTypes.func,
  onDelete: PropTypes.func,
  onLongPress: PropTypes.func,
  onTouchStart: PropTypes.func,
  onTouchMove: PropTypes.func,
  onTouchEnd: PropTypes.func,
  IPCReconfig : PropTypes.object
};

const defaultProps = {
  devId: '',
  online: true,
  isLongPress: false,
  hasOnOffSwitch: false,
  hasLowPower: false,
  icon: 'default',
  statusDesc: '',
  devType: '',
  name: '',
  roomName: '',
  cls: {
    cardIcon: '',
    battery: '',
    switch: '',
    close: '',
  },
  attr: {
    OnOff: '0',
    PowerLow: '0',
  },
  IPCReconfig:{
    text:"",
    func:()=>{}
  },
  onOnOff() { },
  onDelete() { },
  onLongPress() { },
  onTouchStart() { },
  onTouchMove() { },
  onTouchEnd() { },
  
};

class DeviceCard extends Component {
  constructor(props) {
    super(props);
    let isOnOff = 0;
    if (props.attr && props.attr.OnOff !== undefined) {
      isOnOff = (props.attr.OnOff - 0) === 1 ? 1 : 0;
    }
    this.state = {
      OnOff: isOnOff,
      online: props.online,
      attr: props.attr,
    }

    this.moving = false;
    this.isCancelLongPress = false;

    this.handleOnOff = this.handleOnOff.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
    this.handleTouchStart = this.handleTouchStart.bind(this);
    this.handleTouchMove = this.handleTouchMove.bind(this);
    this.handleTouchEnd = this.handleTouchEnd.bind(this);
    this.handleReconfig=this.handleReconfig.bind(this);
  }

  componentDidMount() {
    const { devType, attr, online, hasOnOffSwitch } = this.props;
    if (hasOnOffSwitch) {
      let isOnOff = 0;
      if (attr && attr.OnOff !== undefined) {
        isOnOff = (attr.OnOff - 0) === 1 ? 1 : 0;
      }
      this.setState({
        OnOff: isOnOff
      });
    }
    this.setState({
      online: online
    });
  }

  componentWillReceiveProps(nextProps) {
    const nextAttr = nextProps.attr;
    if (nextAttr && nextAttr.OnOff !== undefined) {
      let isOnOff = 0;
      if (nextAttr && nextAttr.OnOff !== undefined) {
        isOnOff = (nextAttr.OnOff - 0) === 1 ? 1 : 0;
      }
      this.setState({
        OnOff: isOnOff
      });
    }

    if (nextAttr && nextAttr !== undefined) {
      this.setState({
        attr:nextAttr
      });
    }

    this.setState({
      online: nextProps.online
    });
  }

  handleOnOff(e) {
    e.stopPropagation();
    const value = (this.state.OnOff - 0) === 1 ? 0 : 1
    const newAttr = Object.assign(this.state.attr, {});
    newAttr.OnOff = value;
    this.setState({
      OnOff: value,
      attr: newAttr,
    });

    const { devId, onOnOff } = this.props;
    onOnOff(devId, value);
  }

  handleRemove(e) {
    e.stopPropagation();
    const { devId, onDelete } = this.props;
    onDelete(devId);
  }

  handleReconfig(e){
    e.stopPropagation();
    const { IPCReconfig } = this.props;
    IPCReconfig.func();
  }

  handleTouchStart(e) {
    e.stopPropagation();
    if (e.target === this.refs.switchOnOff || e.target === this.refs.delBtn) {
			return;
    }
    if (this.moving) {
			return;
    }
    
    if (this.props.isLongPress) {
      return;
    }

    const { isLongPress } = this.props;
		if (!isLongPress) {
			this.oInterLongPress && clearTimeout(this.oInterLongPress);
			this.oInterLongPress = setTimeout(() => {
        this.oInterLongPress && clearTimeout(this.oInterLongPress);
        this.props.onLongPress(true);
        this._isLongPress = true;
			}, 500);
    }

    const { devId, onTouchStart } = this.props;
    onTouchStart(devId);
  }

  handleTouchMove(e) {
    this.moving = true;
    this.oInterLongPress && clearTimeout(this.oInterLongPress);
    const { devId, onTouchMove } = this.props;
    onTouchMove(devId);
  }

  handleTouchEnd(e) {
    if (e.target === this.refs.switchOnOff || e.target === this.refs.delBtn) {
			return;
		}
		this.oInterLongPress && clearTimeout(this.oInterLongPress);
		if(this.moving) {
			this.moving = false;
			return;
    }
    if (this.isCancelLongPress) {
      this.isCancelLongPress = false;
      return;
    }
    
    const { devId, onTouchEnd } = this.props;
    onTouchEnd(devId);
  }

  render() {
    const { devId, name, roomName, online, icon, statusDesc, devType, isLongPress, cls, hasOnOffSwitch, hasLowPower, securityStatus, onGetStatus,IPCReconfig } = this.props;
    const { OnOff, attr } = this.state;

    const getIconStatus = () => {
      if (!online) {
        return 'offline';
      }
      if (attr && attr.OnOff !== undefined) {
        return (attr.OnOff - 0) === 1 ? 'on' : 'off';
      } else {
        return online ? 'on' : 'off';
      }
    };
    return (
      <div 
        className={`card ${online ? '' : 'disabled'}`}
        onTouchStart={this.handleTouchStart}
				onTouchMove={this.handleTouchMove}
				onTouchEnd={this.handleTouchEnd}
      >
        <div className="card-row fx">
          <div className={`${cls.cardIcon} ${icon} ${getIconStatus()}`} />
          <div className="devices-status">
            {/* <span className="status">{statusDesc}</span> */}
            <span className="status">{onGetStatus({
              devId,
              devType,
              attr,
              online,
              securityStatus
            })}</span>
            {hasOnOffSwitch && <a className={`${cls.switch} ${(OnOff === 0 || !online) ? 'off' : 'on'}`} ref="switchOnOff" onClick={this.handleOnOff}></a>}
          </div>
        </div>
        <div className="card-row name-multi">
          {name}
        </div>
        <div className="card-row fx room">
          <div className="room-name">{roomName}</div>
          {hasLowPower && <div className={cls.battery} />}
        </div>
        {isLongPress ? <a className={cls.close} ref="delBtn" onClick={this.handleRemove}></a> : null}
        {(devType === "IPC" && !online) ? <a className="reconifg" onClick={this.handleReconfig}>{IPCReconfig.text}</a> :"" }
      </div>
    );
  }
}

DeviceCard.propTypes = propTypes;
DeviceCard.defaultProps = defaultProps;
export default DeviceCard;
