// lds-rc/src/commonUI/index.js
import React, { Component } from 'react';
import { BrowserRouter as Router, Route, NavLink, Link, Switch } from 'react-router-dom';
import helper from '../public/helper';
import DialogNormal from '../dialog/normal';
import ToastNormal from '../toast/normal';
import { Style } from './resource';

class CommonUI extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  
  handleShowDialog(type, event) {
  	let dialog = {
  		display: "block",
  		title: "测试标题测试标题测试标题测试标题"
  	};
  	switch(type){
  		case 1:
  			dialog.closable = true;
  			break;
  		case 2:
  			dialog.content = "测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容";
				dialog.buttons = [{
					text: "Okay",
					className: "btnSure",
					handleClick: function(){
						this.hide();
					}
				}];
  			break;
  		case 3:
  			dialog.content = "测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容";
				dialog.buttons = [{
					text: "Okay",
					className: "btnSure",
					handleClick: function(){
						this.hide();
					}
				}, {
					text: "Cancel",
					className: "btnSure",
					handleClick: function(){
						this.hide();
					}
				}];
  			break;
  	}
		this.setState({ dialog });
  }
  
  handleHideDialog() {
  	this.setState({
			dialog: {
    		display: "none"
    	}
		});	
	}
	
  handleShowToast(type, event) {
  	let toast = {
  		display: "block",
  		content: "测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容测试内容"
  	};
  	switch(type){
  		case 1:
  			toast.closable = true;
  			toast.timer = -1;
  			break;
  		case 2:
  			toast.timer = 3000;
  			break;
  	}
		this.setState({ toast });
	}
	
  handleHideToast() {
  	this.setState({
			toast: {
    		display: "none"
    	}
		});	
	}
	
  render() {
  	let _propsDialog = this.state.dialog;
  	let _propsToast = this.state.toast;
  	let {location, history} = this.props;
  	let props = {location, history};
		let myLocation = Object.assign({}, location);

    return (
      <div className="common-ui default">
      	<div>
	        <h2 style={{margin: "20px auto 10px"}}>通用Dialog</h2>
	        <DialogNormal {..._propsDialog} onHide={this.handleHideDialog.bind(this)} /> 
	        <div>
		        <button onClick={this.handleShowDialog.bind(this, 1)}>带标题弹窗</button>
		        <button onClick={this.handleShowDialog.bind(this, 2)}>带内容弹窗</button>
		        <button onClick={this.handleShowDialog.bind(this, 3)}>带标题带内容弹窗</button>
	        </div>
        </div>
      	<div>
	        <h2 style={{margin: "20px auto 10px"}}>通用Toast</h2>
	        <ToastNormal {..._propsToast} onHide={this.handleHideToast.bind(this)} /> 
	        <div>
		        <button onClick={this.handleShowToast.bind(this, 1)}>点击关闭提示窗</button>
		        <button onClick={this.handleShowToast.bind(this, 2)}>计时关闭提示窗</button>
	        </div>
        </div>
        <div>
	        <h2 style={{margin: "20px auto 10px"}}>顶部工具栏</h2>
	        <div></div>
        </div>
        <div>
	        <h2 style={{margin: "20px auto 10px"}}>通用工具栏</h2>
	        <div></div>
        </div>
      </div>
    );
  }
}

export default CommonUI;
