// lds-rc/src/dialog/normal.js
import React, { Component } from 'react';
import PropTypes from 'prop-types'; 
import './default/style.css';
import { Progress} from 'antd-mobile';

export default class DialogNormal extends Component {
		
  constructor(props) {
  	super(props);
  	this.default = {
    	display: "none",
    	closable: false
   };
    this.state = Object.assign({}, this.default, props);
    
    this.handleClickMask = this.handleClickMask.bind(this);
	}
  
  handleClickMask(){
  	if(this.state.closable){
  		this.hide();
  	}
  }
  
  hide(){
    this.props.onHide(this.props);
  }
  
  componentWillReceiveProps(nextProps){
  	for(var key in this.state){
  		delete this.state[key];
  	}
  	this.setState(Object.assign({}, this.default, nextProps));
  }
  
  render() {
  	let _state = this.state;
  	
    return ( 
      <div className="modal" style={{display: _state.display}}>
      	<div className="backdrop" onClick={this.handleClickMask}></div>
      	<div className="dialog">
      		{_state.title &&
	        	<h4 className="title">{_state.title }</h4>
	       	}
      		{_state.content &&
	        	<div className="content" dangerouslySetInnerHTML={{__html: _state.content}}></div>
	       	}
      		{
      			_state.progress && 
      			<div className="progress">
      				<div>updating...<a>{_state.progress}%</a></div>
      				<Progress percent={_state.progress} position="normal" />
      			</div>
      		}
	        {!!_state.buttons && _state.buttons.length &&
		        <div className="operate">
							{_state.buttons.map((item, index) =>
		        		<a key={index} className={item.className} href="javascript:;" onClick={item.handleClick.bind(this)} style={{width: (100 / _state.buttons.length) + "%"}}>
		        			{item.text}
		        		</a>
		        	)}
		        </div>
	       	}
        </div>
      </div>
    );
  }
}

DialogNormal.propTypes = {
  buttons: PropTypes.array,
  closable: PropTypes.bool,
  title: PropTypes.string,
  content: PropTypes.string,
  display: PropTypes.string
}
