// lds-rc/src/toast/normal.js
import React, { Component } from 'react';
import PropTypes from 'prop-types'; 
import './default/style.css';

export default class ToastNormal extends Component {
	
  constructor(props) {
  	super(props);
  	this.default = {
    	display: "none",
    	closable: false,
    	operable: true,
    	timer: 3000
   };
    this.state = Object.assign({}, this.default, props);
    this.handleClickMask = this.handleClickMask.bind(this);
	}
  
  handleClickMask(){
  	if(this.state.closable){
  		this.hide();
  	}
  }
  
  hide(){
    this.props.onHide(this.props);
  }
   
  componentWillUnmount (){
  	this.Timer && clearTimeout(this.Timer);
  }
  
  componentWillReceiveProps(nextProps){
  	for(var key in this.state){
  		delete this.state[key];
  	}
  	let _state = Object.assign({}, this.default, nextProps);
  	if(_state.display === "block" && _state.timer > 0){
  		this.Timer && clearTimeout(this.Timer);
	  	this.Timer = setTimeout(()=>{
		  	this.hide();
	  		this.Timer && clearTimeout(this.Timer);
	  	}, _state.timer)
  	}
  	this.setState(_state);
  }
  
  render() {
  	let _state = this.state;
  	let Toast = (props) => (
  		<div className="toast" onClick={this.handleClickMask} style={{display: _state.display}}>
    		{_state.content &&
        	<div className="content" dangerouslySetInnerHTML={{__html: _state.content}}></div>
       	}
      </div>
  	)
  	/*let styles = {
  		modal: {
				position: "fixed",
				top: 0,
				left: 0,
				zIndex: 1011,
				width: "100%",
				height: "100%"
			},
			backdrop: {
				width: "100%",
				height: "100%",
				background: "rgba(0, 0, 0, 0.6)"
			}
			toast: {
				position: absolute,
				width: "calc(48.92rem * $zoom)",
				top: "50%",
				left: "50%", 
				transform: "translate(-50%, -50%)",
				background: "rgba(0, 0, 0, 0.8)",
				borderRadius: "8px",
				textAlign: "center",
				.content{
					fontSize: "13px",
					lineHeight: "1.44rem",
					margin: "calc(2.67rem * $zoom)",
					wordBreak: "break-word"
				}
			}
  	}*/
  	
  	if(!_state.operable){
  		return (
	    	<div className="modal" style={{display: _state.display}}>
	      	<div className="backdrop" onClick={this.handleClickMask}></div>
		      <Toast />
	      </div>
	    )
  	}else{
  		return <Toast />;
  	}
  }
}

ToastNormal.propTypes = {
  closable: PropTypes.bool,
  operable: PropTypes.bool,
  timer: PropTypes.number,
  content: PropTypes.string,
  display: PropTypes.string
}
